### DISCUSSION TITLE
Task leak troubleshooting

### DISCUSSION THREAD

**Discussion by jertel** [concourse/concourse#6815] [Help & Support] [answered]
Hi all :)  Last week I upgraded my Concourse VMs from 6.7.4 to 7.1.0 and all seemed fine. However, I am having to land/prune/restart my workers every other day now because Concourse can't find a worker available to run new tasks. When I execute `fly -t foo ws -d` I see all my workers have the max number of active tasks, even though there are no pipeline jobs/tasks running. This problem had not occurred until after I upgraded to 7.1.0 so I'm wondering if there's a bug in the 7.x codeline, or if something went wrong with my upgrade. I've not found any other report of this in the forum so I'm leaning towards this being a problem with my environment.

Is there a method I can use to find out what tasks Concourse believes to be still running on each worker? I've looked at `fly -t foo containers` output but it doesn't show me which of those containers are considered 'active'.

**Comment by chenbh** [comment, marked as answer]
If you're using the default `guardian` runtime, then you might be able to use https://github.com/concourse/ft/releases to get runtime metrics on your containers. You'll have to provide it info on how to connect to the db and a single worker, but it's able to inspect them directly and provide more accurate info.

```
$ ft -h
ft is an operator observability tool for concourse

Usage:

Flags:
  -h, --help                          help for ft
      --k8s-namespace string          Kubernetes namespace containing the worker pod to query
      --k8s-pod string                Name of the worker pod to query
      --postgres-ca-cert string       CA cert file location, to verify when connecting to postgres with SSL
      --postgres-client-cert string   Client cert file location, to use when connecting to postgres with SSL
      --postgres-client-key string    Client key file location, to use when connecting to postgres with SSL
      --postgres-database string      The postgres database to connect to (default "atc")
      --postgres-host string          The postgres host to connect to (default "127.0.0.1")
      --postgres-password string      The postgres user's password
      --postgres-port uint16          The postgres port to connect to (default 5432)
      --postgres-sslmode string       Whether or not to use SSL when connecting to postgres (default "disable")
      --postgres-user string          The postgres user to sign in as
      --web-k8s-namespace string      Kubernetes namespace containing the web pod to inpect for connection information
      --web-k8s-pod string            Name of the web pod to inspect for connection information
```

Without specifying the `--k8s-namespace` and `--k8s-pod` flags, it'll assume the worker is on `127.0.0.1:7777` so make sure to `ssh port-forward` first. The output looks something like:
```
workloads                           type   memory    age            handle
main/p/image                        check  164.0 KB  3m44.6561403s  ab6b585d-c666-4527-7f51-f48f7afa31ea
main/p/sleep/1/image                get    18.7 MB   5.8663729s     2e910bf0-16f5-4ffb-4cf2-ee493a61904e
main/p/sleep/1/busy-counting-sleep  task   964.0 KB  4.822841s      d4d59aff-0a20-4dee-722b-32db9b31427f
main/p/sleep/1/image                get    4.6 GB    3m28.8231291s  f22d819a-7e9a-4359-75a0-5914b347f07c
```

**Reply by jertel** [reply]
Thanks for the response. I ran the suggested command but the metrics look normal. There are 5 tasks being output and their ages are less than 3h and memory usage is < 1 MB. I'll keep an eye on it and see if I can find a pattern to the leak. I've noticed I have more of these 'zombied' active tasks when I see more tasks time out during pipeline jobs. But that's just a simple observation and could be unrelated.

**Reply by chenbh** [reply]
Looking at this a bit more, it seems similar to https://github.com/concourse/concourse/issues/6648#issuecomment-801705258. The main difference is that timing out should result in a `failed` build, which shouldn't be affected by that bug.

**Reply by jertel** [reply]
That does look like the issue I'm facing. Last night I reset the workers before my test suites fired off. Four tests timed out overnight. This morning I have 4 'zombied' active tasks across 2 workers. So perhaps timed-out tasks are somehow falling into this situation.

I'll wait for the next release with PR https://github.com/concourse/concourse/pull/6635, and in the meantime continue my work-around of landing, pruning, and restarting workers to get them cleared up.

**Reply by jertel** [reply]
Providing more data on this: One more task has leaked since yesterday, yet no tasks/jobs failed or timed out during that same period. Also, `ft` and `fly workers -d` show no task containers exist. So my theory about this only happening to timed out tasks in my environment is incorrect. 

This doesn't change my plan to wait for the next release. Just some extra data for those interested.

**Reply by jertel** [reply]
Quick follow-up: Release 7.2.0 has resolved this issue. Thanks!
